File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();

        unordered_map<char, int> mp;
        
        int left = 0;
        int maxLength = 0;

        for(int right = 0; right < n; right++){            
            
            if(mp.find(s[right]) != mp.end()){
                left = max(left, mp[s[right]] + 1);
            }

            mp[s[right]] = right;
            maxLength = max(maxLength, right - left + 1);
        }

        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();

        Set<Character> st = new HashSet<>();

        int maxLen = 0;
        int a = 0;
        int b = 0;

        while(b < n){

            if(st.contains(s.charAt(b))){
                st.remove(s.charAt(a));
                a++;
            }else{
                st.add(s.charAt(b));

                maxLen = Math.max(maxLen, b - a + 1);

                b++;
            }
        }

        return maxLen;
    }
}